[Serializer] Add ChainNameConverter

jeromegamez opened the pull request:
| Q | A |
| --- | --- |
| Branch | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR | symfony/symfony-docs/pull/6374 |

The `ChainNameConverter` allows to process property names with multiple converters. This can be useful e.g. when you have to CamelCase underscored property names, and then need to rename them by custom rules.

My current use case is that I receive serialized data with (some, but not all) german snake cased property names (it's a 3rd party API :smile:) which I need to camelcase _and_ translate to english, so that e.g. a property named `ist_archiviert` would first be converted to `istArchiviert` and then translated into `isArchived` by a second converter. Properties that are already in english language would just be camelcased, and already valid property names would remain unchanged.

An usage example (with the `OrgPrefixNameConverter` from the docs):

``` php
use Symfony\Component\Serializer\NameConverter\CamelCaseToSnakeCaseNameConverter;
use Symfony\Component\Serializer\NameConverter\ChainNameConverter;

$camelCaseNameConverter = new CamelCaseToSnakeCaseNameConverter();
$orgPrefixNameConverter = new OrgPrefixNameConverter();

$nameConverter = new ChainNameConverter(array($camelCaseNameConverter, $orgPrefixNameConverter));
$normalizer = new ObjectNormalizer(null, $nameConverter);

:octocat: 
```

xabbuh commented on src/Symfony/Component/Serializer/NameConverter/ChainNameConverter.php:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * @var NameConverterInterface[]
+     */
+    protected $converters = array();
```
should be `private`

jeromegamez replied:
Changed (I prefer `private`s, too, but I took the ChainLoader as an example - which perhaps wasn't the best idea, because it's so old already :))

xabbuh commented on src/Symfony/Component/Serializer/NameConverter/ChainNameConverter.php:
```diff
@@ -0,0 +1,69 @@
+    {
+        foreach ($converters as $converter) {
+            $this->addConverter($converter);
+        }
```
just use `$this->converters = $converters`

jeromegamez replied:
If someone adds a class that doesn't implement `NameConverterInterface`, they would get an error later - would this be acceptable?

xabbuh replied:
Let's change the param type in the docblock to `NameConverterInterface[]`. Then it's the user's responsibility to follow the contract that we make here.

jeromegamez replied:
It's changed - I also removed the `addConverter` method, as it's not needed anymore, and it's unlikely that a user needs to add a converter when they already could have added it in the constructor anyways.

jeromegamez commented:
By the way: as this - if merged - is going to end up in Symfony 3.x, which has a requirement of PHP >= 5.5.9 - would I be allowed to use `NameConverterInterface::class` instead of `'Symfony\Component\Serializer\NameConverter\NameConverterInterface'` in the Unit Tests, and `[...]` instead of `array(...)`?

xabbuh commented:
@jeromegamez yes

jeromegamez commented:
Perfect, then I will do that!

dunglas commented:
To be exact, yes for `::class` (it's even enforced for new code), but no for the short array syntax for coherence with the existing code base and simplify the merge between branches process.

dunglas commented on src/Symfony/Component/Serializer/NameConverter/ChainNameConverter.php:
```diff
@@ -0,0 +1,57 @@
+class ChainNameConverter implements NameConverterInterface
+{
+    /**
+     * @var NameConverterInterface[]
```
This PHPDoc can be removed (the one in the constructor is enough).

jeromegamez replied:
Done!

dunglas commented:
Why don't you decorate the `CamelCaseToSnakeCaseNameConverter` with your own decorator instead on chaining them?

jeromegamez commented:
@dunglas That's what I am currently doing in my project, I just thought it would be nice to have a chain converter in case one has to to use two or more converters in different serializers - if one of the converters is not needed anymore, one needs just to remove it from the chain.

I realize a `ChainNameConverter` is a convenience class and it might not be used by a lot of people, I just thought it's lightweight and simple enough to be included, similarly to the `ChainDecoder`.

jeromegamez committed:
[Serializer] Add ChainNameConverter

The ChainNameConverter allows to process property names with multiple converters.

dunglas commented:
ping @symfony/deciders

fabpot commented:
My fear is that depending on the order of name converter registrations, the result can be different. If that's not a too common use case and because this is possible by using decoration, I'm not sure we should add this to core. @dunglas Your call :)
